sql/core/src/main/scala/org/apache/spark/sql/package.scala

```diff
@@ -73,4 +76,43 @@ package object sql {
    * with rebasing.
    */
   private[sql] val SPARK_LEGACY_INT96_METADATA_KEY = "org.apache.spark.legacyINT96"
+
+  /**
+   * This helper function captures the Spark API and its call site in the user code from the current
+   * stacktrace.
+   *
+   * As adding `withOrigin` explicitly to all Spark API definition would be a huge change,
+   * `withOrigin` is used only at certain places where all API implementation surely pass through
+   * and the current stacktrace is filtered to the point where first Spark API code is invoked from
+   * the user code.
+   *
+   * As there might be multiple nested `withOrigin` calls (e.g. any Spark API implementations can
+   * invoke other APIs) only the first `withOrigin` is captured because that is closer to the user
+   * code.
+   *
+   * @param framesToDrop the number of stack frames we can surely drop before searching for the user
+   *                     code
+   * @param f the function that can use the origin
+   * @return the result of `f`
+   */
+  private[sql] def withOrigin[T](framesToDrop: Int = 0)(f: => T): T = {
+    if (CurrentOrigin.get.stackTrace.isDefined) {
+      f
+    } else {
+      val st = Thread.currentThread().getStackTrace
+      var i = framesToDrop + 3
+      while (sparkCode(st(i))) i += 1
```

Comment:
We set `framesToDrop = 1` in a few places:
- `Column.fn`
- `withExpr`
- `repartitionByExpression`
- `repartitionByRange`
- `withAggregateFunction`
- `createLambda`

So, there are 2 options either
- the function `sparkCode` doesn't work properly, and we skip 1 frame forcibly
- or a premature optimization.

I will check that after all tests passed eventually.

Reply:
I wated to use it only for optimization, at certrain places we simply know that at least how many frames deep we are in Spark's code. `sparkCode()` uses regex so can be a bit slow...

Reply:
> sparkCode() uses regex so it can be a bit slow.

Shouldn't be so slow, I think. Especially, just one pattern match. I'll remove the optimization so far.